library/math/dual.lisp

```diff
@@ -0,0 +1,65 @@
+;;;; dual.lisp
+;;;;
+;;;; dual numbers to enable automatic differentiation
+
+(coalton-library/utils:defstdlib-package #:coalton-library/math/dual
+  (:use
+   #:coalton
+   #:coalton-library/builtin
+   #:coalton-library/classes
+   #:coalton-library/functions
+   #:coalton-library/math/arith)
+  (:export
+   #:Dual
+   #:dual-real
+   #:dual-dual
+   #:sub-duals
+   #:add-duals
+   #:mul-duals))
+
+(in-package #:coalton-library/math/dual)
+
+(coalton-toplevel
+
+ (define-type *Real*
+     (*Real* Integer))
+
+ (define-type *Dual*
+     (*Dual* Integer))
+ 
+ (define-type Dual
+     (Dual *Real* *Dual*))
+ 
+ (declare dual-real (Dual -> Integer))
```

Comment:
i'd name it `real-part`

with the above change the function would have type `Dual :t -> :t`